ui/components/app/currency-input/currency-input.js

```diff
@@ -41,84 +45,101 @@ export default function CurrencyInput({
   const isOriginalNativeSymbol = useIsOriginalNativeTokenSymbol(
     chainId,
     ticker,
     type,
   );
-  const hideSecondary = !showFiat;
-  const primarySuffix = preferredCurrency || EtherDenomination.ETH;
-  const secondarySuffix = secondaryCurrency.toUpperCase();
 
-  const [isSwapped, setSwapped] = useState(false);
-  const [newHexValue, setNewHexValue] = useState(hexValue);
-  const [shouldDisplayFiat, setShouldDisplayFiat] = useState(featureSecondary);
-  const shouldUseFiat = hideSecondary ? false : Boolean(shouldDisplayFiat);
-
-  const getDecimalValue = () => {
-    const decimalValueString = shouldUseFiat
-      ? getValueFromWeiHex({
-          value: hexValue,
-          toCurrency: secondaryCurrency,
-          conversionRate,
-          numberOfDecimals: 2,
-        })
-      : getValueFromWeiHex({
-          value: hexValue,
-          toCurrency: EtherDenomination.ETH,
-          numberOfDecimals: 8,
-        });
-
-    return Number(decimalValueString) || 0;
-  };
+  const tokenToFiatConversionRate = useTokenExchangeRate(asset?.address);
 
-  const initialDecimalValue = hexValue ? getDecimalValue() : 0;
+  const processNewDecimalValue = useProcessNewDecimalValue(
+    assetDecimals,
+    isTokenPrimary,
+    tokenToFiatConversionRate,
+  );
 
   const swap = async () => {
     await onPreferenceToggle();
-    setSwapped(!isSwapped);
-    setShouldDisplayFiat(!shouldDisplayFiat);
+    setShouldDisplayFiatIfAvailable(!shouldDisplayFiatIfAvailable);
   };
 
+  // if the conversion rate is undefined, do not allow a fiat input
+  useEffect(() => {
+    if (isTokenPrimary) {
+      return;
+    }
+
+    if (!tokenToFiatConversionRate) {
+      onPreferenceToggle();
+      setShouldDisplayFiatIfAvailable(false);
+    }
+  }, [tokenToFiatConversionRate, isTokenPrimary, onPreferenceToggle]);
+
   const handleChange = (newDecimalValue) => {
-    const hexValueNew = shouldUseFiat
-      ? getWeiHexFromDecimalValue({
-          value: newDecimalValue,
-          fromCurrency: secondaryCurrency,
-          conversionRate,
-          invertConversionRate: true,
-        })
-      : getWeiHexFromDecimalValue({
-          value: newDecimalValue,
-          fromCurrency: EtherDenomination.ETH,
-          fromDenomination: EtherDenomination.ETH,
-          conversionRate,
-        });
-
-    setNewHexValue(hexValueNew);
-    onChange(hexValueNew);
-    setSwapped(!isSwapped);
+    const { newTokenDecimalValue, newFiatDecimalValue } =
+      processNewDecimalValue(newDecimalValue);
+    setTokenDecimalValue(newTokenDecimalValue);
+    setFiatDecimalValue(newFiatDecimalValue);
+
+    onChange(
+      new Numeric(newTokenDecimalValue, 10)
+        .times(Math.pow(10, assetDecimals), 10)
+        .toPrefixedHexString(),
+    );
   };
 
+  // reset form when token is changed
   useEffect(() => {
-    setNewHexValue(hexValue);
-  }, [hexValue]);
+    setTokenDecimalValue('0');
+    setFiatDecimalValue('0');
+  }, [asset?.address]);
```

Comment:
Does onChange need to be called as well to keep the draftTransaction up-to-date? I'm unclear on whether some other part of the code resets the draft tx amount when the token changes (I've seen it happen sometimes before this fix went in). It would be easier to trace these state updates if the tx amount was explicitly reset/updated in content.tsx:handleSelectToken instead

Reply:
It does get reset in `updateSendAsset`, but the logic's a bit convoluted, so I added it per your suggestion.

Reply:
Going back on this since it's not 100% necessary and breaks a bunch of tests